hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/has/isauthorizedraw/IsAuthorizedRawCall.java

```diff
@@ -0,0 +1,266 @@
+
+        // Fail fast if user didn't supply monstrous amount of gas
+        final var availableGas = frame.getRemainingGas();
+        if (availableGas < HARDCODED_GAS_REQUIREMENT_GAS) return bail.apply(INSUFFICIENT_GAS);
+
+        // Figure out what kind of signature we're dealing with thus what kind of account+key we need
+        final var signatureType =
+                switch (signature.length) {
+                    case 65 -> SignatureType.EC;
+                    case 64 -> SignatureType.ED;
+                    default -> SignatureType.Invalid;
+                };
+        if (signatureType == SignatureType.Invalid)
+            return bail.apply(FAIL_INVALID /* should be: "invalid argument to precompile" */);
+
+        // Validate parameters according to signature type
+        if (!switch (signatureType) {
+            case EC -> messageHash.length == 32;
+            case ED -> true;
+            default -> throw new IllegalStateException("Unexpected value: " + signatureType);
+        }) return bail.apply(FAIL_INVALID /* should be: "invalid argument to precompile */);
+
+        // Gotta have an account that the given address is an alias for
+        final var accountNum = accountNumberForEvmReference(address, nativeOperations());
+        if (!isValidAccount(accountNum, signatureType)) return bail.apply(INVALID_ACCOUNT_ID);
+        final var account = requireNonNull(enhancement.nativeOperations().getAccount(accountNum));
+
+        // If ED then require a key on the account
+        final Optional<Key> key;
+        if (signatureType == SignatureType.ED) {
+            key = Optional.ofNullable(account.key());
+            if (key.isEmpty()) return bail.apply(FAIL_INVALID /* should be: "account must have key" */);
+        } else key = Optional.empty();
+
+        // Key must be simple (for isAuthorizedRaw)
+        if (key.isPresent()) {
+            final Key ky = key.get();
+            final boolean keyIsSimple = !ky.hasKeyList() && !ky.hasThresholdKey();
+            if (!keyIsSimple) return bail.apply(FAIL_INVALID /* should be: "account key must be simple" */);
+        }
+
+        // Key must match signature type
+        if (key.isPresent()) {
+            if (!switch (signatureType) {
+                case ED -> key.get().hasEd25519();
+                case EC -> false;
+                default -> false;
+            }) return bail.apply(INVALID_SIGNATURE_TYPE_MISMATCHING_KEY);
+        }
+
+        // Finally: Do the signature validation we came here for!
+        final boolean authorized =
+                switch (signatureType) {
+                    case EC -> validateEcSignature(address, frame);
+                    case ED -> validateEdSignature(account, key.get());
+                    default -> false;
+                };
+
+        final var result = gasOnly(
+                successResult(encodedAuthorizationOutput(authorized), HARDCODED_GAS_REQUIREMENT_GAS), SUCCESS, true);
+        return result;
```

Comment:
```suggestion
        return gasOnly(
                successResult(encodedAuthorizationOutput(authorized), HARDCODED_GAS_REQUIREMENT_GAS), SUCCESS, true);
```

Reply:
We could simplify this :)

Reply:
I know the simplification but generally I prefer to do it this way - via an intermediate named local variable before returning the result - to make debugging now (and in the future!) easier.  You can see the result to be returned _before_ you actually return from the function, when in the debugger.

I've found this to be important many times when debugging.  And I like to leave it in the code because when you're debugging against a release (for example) you don't have the luxury of being able to change the code at that time.  So ... I'll leave it like this.

Reply:
Oh, got it. Yeah, I agree. 
Just a small suggestion then: What do you think about changing it to boolean, as I feel like var is a little bit of an "encoding" for primitive types?

Reply:
Oh, yes. I agree :)